src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -258,6 +261,148 @@ pub trait TimestampProvider {
+
+            // First, we have validity constraints from the `id_bundle` argument which indicates
+            // which collections we are reading from.
+            // TODO: Refine the detail about which identifiers are binding and which are not.
+            let since = self.least_valid_read(&read_holds);
+            let storage = id_bundle.storage_ids.iter().cloned().collect();
+            constraints
+                .lower
+                .push((since.clone(), Reason::StorageInput(storage)));
+            let compute = id_bundle
+                .compute_ids
+                .iter()
+                .flat_map(|(key, ids)| ids.iter().map(|id| (*key, *id)))
+                .collect();
+            constraints
+                .lower
+                .push((since.clone(), Reason::ComputeInput(compute)));
+
+            // The query's `when` may indicates a specific timestamp we must advance to, or a specific value we must use.
+            if let Some(timestamp) = when.advance_to_timestamp() {
+                let catalog_state = self.catalog_state();
+                let ts = Coordinator::evaluate_when(catalog_state, timestamp, session)?;
+                constraints
+                    .lower
+                    .push((Antichain::from_elem(ts), Reason::QueryAsOf));
+                // If the query is at a specific timestamp, we must introduce an upper bound as well.
+                if let QueryWhen::AtTimestamp(_) = when {
+                    constraints
+                        .upper
+                        .push((Antichain::from_elem(ts), Reason::QueryAsOf));
+                }
+            }
+
+            // The specification of an `oracle_read_ts` may indicates that we must advance to it,
+            // except in one isolation mode, or if `when` does not indicate that we should.
+            // At the moment, only `QueryWhen::FreshestTableWrite` indicates that we should.
+            // TODO: Should this just depend on the isolation level?
+            if let Some(timestamp) = &oracle_read_ts {
+                if isolation_level != &IsolationLevel::StrongSessionSerializable
+                    || when.must_advance_to_timeline_ts()
+                {
+                    constraints.lower.push((
+                        Antichain::from_elem(*timestamp),
+                        Reason::IsolationLevel(*isolation_level),
+                    ));
+                }
+            }
+
+            // If a real time recency timestamp is supplied, we must advance to it.
+            if let Some(real_time_recency_ts) = real_time_recency_ts {
+                assert!(
+                    session.vars().real_time_recency()
+                        && isolation_level == &IsolationLevel::StrictSerializable,
+                    "real time recency timestamp should only be supplied when real time recency \
+                                is enabled and the isolation level is strict serializable"
+                );
```

Comment:
Given the concept of a "shadow/validation rollout", `assert`-ing in this code seems bad since it can panic environmentd?

Could we move this logic to a separate error-able function and when validation is enabled, log the possible error?

Reply:
Looks like we're still `assert!`-ing here?